WIP: Custom bundle downloaders

tsandall opened the pull request:
<!--

Thanks for submitting a PR to OPA!

Before pressing 'Create pull request' please read the checklist below.

* All code changes should be accompanied with tests. If you are not
modifying any tests, just provide a short explanation of why updates
to tests are not necessary. In addition to helping catch bugs, tests
are extremely helpful in providing _context_ that explains how your
changes can be used.

* All changes to public APIs **must** be accompanied with
docs. Examples of public APIs include built-in functions,
config fields, and of course, exported Go types/functions/constants/etc.

* Commit messages should explain _why_ you made the changes, not what
you changed. Use active voice. Keep the subject line under 50
characters or so.

* All commits must be signed off by the author. If you are not
familiar with signing off, see CONTRIBUTING.md below.

For more information on contributing to OPA see:

* [CONTRIBUTING.md](https://github.com/open-policy-agent/opa/blob/master/CONTRIBUTING.md)
  for high-level contribution guidelines.

* [DEVELOPMENT.md](https://github.com/open-policy-agent/opa/blob/master/docs/devel/DEVELOPMENT.md)
  for development workflow and environment setup.

-->

patrick-east commented on download/download.go:
```diff
@@ -26,6 +26,21 @@ const (
+
+// Interface can be implemented to download via arbitrary transports/formats.
+type Interface interface {
+	WithLogConfig(LogConfig) Interface                    // set log configuration parameters
```
We can probably omit the `WithLogConfig` and `LogConfig` part of the interface, a quick check with the github search doesn't show the older `WithLogAttrs` being used anywhere.

I'd vote to keep it on the standard http downloader so that anyone using one directly could still use it, for backwards compatibility, but leave it out of the new interface until we need it (unless we already do and I'm missing it)

tsandall replied:
I think the log attributes/configuration are useful for correlating rest client debug messages with the plugin that owns the rest client. I also thought about removing it but while debugging a recent issue, I noticed we're not passing the log attributes down and it was a bit annoying to tell which plugin was making the call.

patrick-east replied:
>  I noticed we're not passing the log attributes down and it was a bit annoying to tell which plugin was making the call.

Ahh yea, if we fix that I'm definitely on board with keeping it.

tsandall replied:
It's easy enough to fix and this would be a bit more future proof (struct instead of slice). If you're onboard, I'll fixup the PR and merge then?

patrick-east replied:
Yea, lets do it.

tsandall commented:
@patrick-east while we're defining the interface, any thoughts on ClearCache()? Also, I'm tempted to do discovery at the same time to work through any design issues. WDYT?

patrick-east commented:
> while we're defining the interface, any thoughts on ClearCache()? Also, I'm tempted to do discovery at the same time to work through any design issues. WDYT?

I've been wondering about that. I like having that kind of control in the clients, but after letting it soak for a little while I'm wonder if maybe clients would be better served with an even more generic `reset()` or `restart()` or something.

I guess my thinking is that the client would want to clear any cached data, but likely any current or saved state should be wiped. Maybe even aborting some in-progress request processing or something. As the implementor of some new downloader type it might be easier to have the logic centralized in the generic one, and for a client it's a bigger and higher level hammer to hit their downloader with when something goes wrong. They may or may not care whether the cache is cleared (or even know what is cached), they just want it to start fresh.

tsandall commented:
@patrick-east another alternative is to have the callback return an indication to the downloader about what happened (or maybe just an error.) The downloader can do whatever it wants with that. That doesn't let cache get cleared asynchronously but it doesn't look like we're using it that way today. One benefit to returning an error is that the downloader would know which consumer had an issue in a scenario where there are multiple callbacks. That said, I'm not sure the async or multiple use cases are likely...the cache clearing logic only manifested because we split the downloader from the bundle plugin and the decision about whether to clear cannot be made by the downloader alone. Any thoughts?

patrick-east commented:
> another alternative is to have the callback return an indication to the downloader about what happened (or maybe just an error.) The downloader can do whatever it wants with that.

I like that, would maybe pair nicely with some way to supply caching config (maybe just specific to a particular downloader implementation). I could see having a boolean option for like "cache on callback error" or something.

> That said, I'm not sure the async or multiple use cases are likely...the cache clearing logic only manifested because we split the downloader from the bundle plugin and the decision about whether to clear cannot be made by the downloader alone.

Agreed, I think the async option is potentially useful for like the plugin being reconfigured, but I think right now we just stop/start/replace them as needed. We probably don't need another API to assist with that.

tsandall commented:
@patrick-east PTAL at the latest changes. I've replaced ClearCache with an error return value and removed the LogConfig stuff. I just realized that adding support to discovery will require defining another factory interface (unless we do it in the downloader package.)

patrick-east commented on download/download.go:
```diff
@@ -26,6 +26,13 @@ const (
 )
 
+// Interface can be implemented to download via arbitrary transports/formats.
+type Interface interface {
```
🤔 just thinking out loud a little bit, `download.Interface` is a little.. i dunno.. maybe too close to just `interface{}` or something. Any thoughts on calling it something more like `download.Client` or `download.Transport` or something?

The first time I read through the PR I was just like "ah yea, its the download interface" but now I'm having some second thoughts on it.

patrick-east commented on download/download.go:
```diff
@@ -61,23 +67,11 @@ func New(config Config, client rest.Client, path string) *Downloader {
 
 // WithCallback registers a function f to be called when download updates occur.
-func (d *Downloader) WithCallback(f func(context.Context, Update)) *Downloader {
+func (d *Downloader) WithCallback(f func(context.Context, Update) error) Interface {
```
Is it going to be OK for us to change the public API? I was anticipating a different name for the new one that had an error response.

I guess if we assume no one else is using the downloader its maybe not a concern.

tsandall commented:
Closing this for now. We can revisit when custom downloading is really needed.
